### ISSUE TITLE
Why ViewModel has no access to presenter? How can I access to individual label?

### ISSUE THREAD

**Issue by shungz89** [badoo/Chatto#523]
This is ridiculous, I have spent TWO WEEKS because I'm stuck with this framework of not being able to customize many component. Such as DYNAMICALLY changing the textmessagebubble based on whether the incoming message belongs to certain parties.

Why do you make such framework which is so hard to access individual component?
If is so, then why can't you make simple stuff like that customizable? 

Even changing the AvatarImageView I had to spend 2-3 hours to get it done because the Demo sample do not provide DIRECT access to it, you actually have to copy and paste some class and then modify it.

**Comment by AntonPalich** [comment]
@shungz89 Imagine, that incoming messages from one party should have blue bubble and borders, incoming messages from the second party should have red bubble and no borders and incoming messages from the third party one shouldn't have bubble and borders at all. Instead of trying to use the same type of message to display messages from certain parties you can try to use different type of messages for every party involved. At this point all the difference will be at configuration point. So, in this case you may have three message types with three presenter builders that only differs in styles that were used.

Let me show you an example:

```swift
override func createPresenterBuilders() -> [ChatItemType: [ChatItemPresenterBuilderProtocol]] {

    let textMessagePresenter = TextMessagePresenterBuilder(
        viewModelBuilder: DemoTextMessageViewModelBuilder(),
        interactionHandler: DemoTextMessageHandler(baseHandler: self.baseMessageHandler)
    )
    textMessagePresenter.baseMessageStyle = BaseMessageCollectionViewCellDefaultStyle(
        colors: BaseMessageCollectionViewCellDefaultStyle.Colors(
            incoming: .blue,
            outgoing: .blue
        )
    )

    let anotherTextMessagePresenter = TextMessagePresenterBuilder(
        viewModelBuilder: DemoTextMessageViewModelBuilder(),
        interactionHandler: DemoTextMessageHandler(baseHandler: self.baseMessageHandler)
    )
    anotherTextMessagePresenter.baseMessageStyle = BaseMessageCollectionViewCellDefaultStyle(
        colors: BaseMessageCollectionViewCellDefaultStyle.Colors(
            incoming: .red,
            outgoing: .red
        )
    )

    return [
        DemoTextMessageModel.oneChatItemType: [textMessagePresenter],
        DemoTextMessageModel.anotherChatItemType: [anotherTextMessagePresenter],
    ]
}
```

In this example I've registered two message types: one of them displays text with blue bubbles, another one displays text with red bubbles. Your model is responsible for defining its type taking into account certain parties. This is the main idea. As you can see, you can do dynamic behaviour in different ways.

> Why do you make such framework which is so hard to access individual component?
If is so, then why can't you make simple stuff like that customizable?

There should be strict layer separation that is a good engineering practice and gives your lots of benefits. A **View** shouldn't know about anything and provide a way to customize itself through **ViewModel**. In case of this framework **ViewModel** doesn't have a business logic and it's stores data for a **View**. A **Presenter** is responsible for cell configuration and having business logic. It configures a cell through setting up a **ViewModel** with specific values based on your business logic. Answering your questions from the title, you shouldn't try to access a **Presenter** from a **ViewModel** because it inverse relationships between them, also you shouldn't access UI elements from a **Presenter** because your jump between layers in this case. Instead, you should  cusomize your **ViewModel** in a **Presenter** to configure a **View**.

**Comment by shungz89** [comment]
Okay, is a good idea on the creating two items type, but how do you create two types of item?

For my requirements, I would need to access individual message label because I need to change it into html attributed text. Without access to the UILabel it is not easily doable.

**Comment by AntonPalich** [comment]
@shungz89 

> Okay, is a good idea on the creating two items type, but how do you create two types of item?

Imagine you have a model class in your project that represents a message. It must confirm to the most basic `ChatItemProtocol` that defines a `type` variable. So, value of this property defines a type of the message. This value can be dynamic and depend on other properties of you message. For example:

```swift
class Message: ChatItemProtocol {
    let isSystemMessage: Bool

    // MARK: - ChatItemProtocol
    var type: String {
        if self.isSystemMessage {
            return Message.oneType
        } else {
            return Message.anotherType
        }
    }

    static var oneType: String = "oneType"
    static var anotherType: String = "anotherType"
}
```

After that you can register two different presenter builders for `Message.oneType` and `Message.anotherType`. This will allow you to display messages with different types using the same model.

> For my requirements, I would need to access individual message label because I need to change it into html attributed text. Without access to the UILabel it is not easily doable.

The issue with attributed text was described here: https://github.com/badoo/Chatto/issues/462

TLDR: Currently, attributed strings are not supported, but you can add support in your project on your own by copy-pasting and modifying classes responsible for displaying text messages (https://github.com/badoo/Chatto/issues/462#issuecomment-372590898)
